code/game/machinery/howitzer.dm

```diff
@@ -0,0 +1,532 @@
+	pixel_x = -16
+
+	/**
+	 * List of allowed ammo types that this howitzer will accept
+	 */
+	var/allowed_ammo_types = list(/obj/item/ammo_casing)
+
+	///The targeted horizontal angle requested to this howitzer
+	var/horizontal_angle = 0
+
+	///The targeted vertical angle requested to this howitzer
+	var/vertical_angle = 0
+
+	///The deciseconds it takes for the howitzer to rotate by one degree, horizontally
+	var/decisecond_per_degree_horizontal_rotation = 1
+
+	///The deciseconds it takes for the howitzer to change the barrel elevation by one degree, vertically
+	var/decisecond_per_degree_vertical_rotation = 1.5
+
+	///The shot that is currently loaded into this howitzer
+	var/obj/item/ammo_casing/howitzer/loaded_shot
+
+	///The sound to play when the howitzer fire, a string
+	var/fire_sound = 'sound/machines/howitzer/fire.ogg'
+
+	///If the howitzer is ready to fire, handled internally
+	VAR_PRIVATE/ready_to_fire = FALSE
+
+	///A list of `/obj/item/howitzer_pellet` that are currently loaded into this howitzer
+	var/list/obj/item/howitzer_pellet/loaded_howitzer_pellets = list()
+
+
+	/* START ROTATION INTERNALS */
+
+	///The sound this howitzer will emit while rotating
+	var/rotation_looping_sound_type = /datum/looping_sound/clanking
+
+	///The looping sound used during the howitzer rotation
+	VAR_PRIVATE/datum/looping_sound/rotation_looping_sound
+
+	///The timerid of the rotation
+	VAR_PRIVATE/rotation_timerid
+
+	/* END ROTATION INTERNALS */
+
+/obj/machinery/howitzer/Initialize(mapload, d, populate_components, is_internal)
+	. = ..()
+	if(!anchored)
+		icon_state = "howitzer"
+
+	set_dir(NORTH)
+
+/obj/machinery/howitzer/attack_hand(mob/user)
+	if(istype(user, /mob/living/carbon/human))
+		var/mob/living/carbon/human/user_human = user
+
+		if(loaded_shot)
+			//Remove the pellets first, the projectile last
+			if(length(loaded_howitzer_pellets))
+				var/obj/item/howitzer_pellet/pellet = loaded_howitzer_pellets[length(loaded_howitzer_pellets)]
+				pellet.forceMove(src)
+				user_human.put_in_hands(pellet)
+				loaded_howitzer_pellets -= pellet
+
+			else
+				loaded_shot.forceMove(src)
+				user_human.put_in_hands(loaded_shot)
+				loaded_shot = null
+			return
+
+		ui_interact(user)
+
+/obj/machinery/howitzer/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Howitzer", ui_x=500, ui_y=380)
+		ui.open()
+
+/obj/machinery/howitzer/ui_data(mob/user)
+	var/list/data = list()
+	data["loaded_shot"] = loaded_shot ? TRUE : FALSE
+	data["horizontal_angle"] = horizontal_angle
+	data["vertical_angle"] = vertical_angle
+
+	return data
+
+/obj/machinery/howitzer/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	if(.)
+		return
+
+	switch(action)
+		if("set_horizontal_angle")
+			change_horizontal_angle(params["horizontal_angle"])
+		if("set_vertical_angle")
+			change_vertical_angle(params["vertical_angle"])
+
+		if("fire")
+			fire(ui.user)
```

Comment:
these don't return TRUE, does that mean the UI doesn't need to update?

Reply:
Well it doesn't update immediately, but it wouldn't have the updated value anyways, it gets updated at the next time the tgui refresh processing runs